### ISSUE TITLE
[tune] A few differences between the Hyperband implementation and the original paper

### ISSUE THREAD

**Issue by haoyangz** [ray-project/ray#1624]
Comparing to the original [post](https://people.eecs.berkeley.edu/~kjamieson/hyperband.html), I spotted a few differences (not sure if there are others):

- [s_max](https://github.com/richardliaw/ray/blob/8b85e7bc4aa96af8fb56485702e3490c8dfd2a0d/python/ray/tune/hyperband.py#L75) is currently a fixed value (=5). It should be calculated from max_t (with the definition in the original post) and eta (using `int(log(max_iter)/log(eta))`)

- [n_0](https://github.com/richardliaw/ray/blob/8b85e7bc4aa96af8fb56485702e3490c8dfd2a0d/python/ray/tune/hyperband.py#L79) is calculated differently. The original post rounded (s_max+1)/(s+1) whereas the current implementation doesn't, leading to [trial sizes](https://github.com/ray-project/ray/blob/31fefa20b746e2ae8d9e9af3c72e61112bcb9725/python/ray/tune/test/trial_scheduler_test.py#L193-L197) that are different from the original post (always an exponent of 3).

**Comment by haoyangz** [comment]
@richardliaw

**Comment by richardliaw** [comment]
Nice! It would probably be good to clarify all of these points in the implementation details. 

For context, our [original implementation of Hyperband](https://github.com/richardliaw/ray/blob/f945ef479fe13dc9012cfe4511cfd0efda7686c0/python/ray/tune/hyperband.py) actually followed the paper, but after using it for a while, we realized that the given parameters led to many unintuitive and impractical bracket configurations. 

1. Both `s_max` (representing the number of brackets - 1) and `eta`, representing the downsampling rate, are fixed.  In many practical settings, you may have `R >= 200`. For simplicity, assume `eta = 3`. Between `R = 200` and `R = 1000`, you end up with a huge range of the number trials needed to fill up all brackets - 
![screenshot 2018-02-27 11 33 38](https://user-images.githubusercontent.com/4529381/36750419-4f3ebc9e-1bb2-11e8-9d0b-1f6c491efdfd.png)
Now, assume `R = 300`. If you want to find a better `eta`, you realize also that the configurations generated aren't very intuitive:
![screenshot 2018-02-27 11 36 55](https://user-images.githubusercontent.com/4529381/36750529-9baf7a32-1bb2-11e8-856f-772bbcd67ac9.png)
We figured that it would be a lot easier for majority of people to use if we fixed the parameters to be the same as the example given in the paper and only exposed `max_t`. I actually wouldn't worry too much about these deviations from theory, because HyperBand is actually quite inefficient in implementation (both in our experience and [noted by the authors](https://openreview.net/forum?id=S1Y7OOlRZ), who propose a fix that [we will soon provide](https://github.com/ray-project/ray/pull/1595)).

2. The example in the post to calculate `n_0` is actually a little different than the algorithm given in the paper. 
Screenshot from the original paper: 
![screenshot 2018-02-27 10 55 09](https://user-images.githubusercontent.com/4529381/36751005-c7bb4664-1bb3-11e8-87b5-f76dab8237b3.png)

In fact, not rounding the calculation inside actually provides fairer resource allocation to all of the brackets.
![screenshot 2018-02-27 11 54 21](https://user-images.githubusercontent.com/4529381/36751413-f9c2a5b6-1bb4-11e8-83d5-bafda8e6aa1d.png)

3. There are also implementation specific details like how trials are placed into brackets which are not covered in the paper. Also, the recent PR #1620 aims to cut trials after `max_t`, which might actually not be the semantic definition in the paper, but makes much more sense to the end user.

tldr implementation details may deviate slightly from theory but are focused on increasing usability

**Comment by ericl** [comment]
@richardliaw it could be useful to move that info into a doc, e.g. in the hyperband page of https://github.com/ray-project/ray/pull/1642

**Comment by richardliaw** [comment]
#1642 is merged now; closing this. Feel free to open if anything else comes up!
